### ISSUE TITLE
Preview API

### ISSUE THREAD

**Issue by tobiasKaminsky** [nextcloud/server#3923]
Currently we have a thumbnail API and a preview API, provided by gallery app.
In order to provide proper previews for the clients it would be nice to have a reliable API within the core part.

The API should do 
- give previews on "any" size
- obey aspect ratio
- give correct landscape/portrait orientation
 

@rullzer (cause we talked about it)
@oparoz as your gallery API is doing all those requirements already :)

**Comment by oparoz** [comment]
Yes, the Preview API should be its own app and come with a provider manager and an API. The thumbnail API has too many issues.

**Comment by rullzer** [comment]
So few things:

> give previews on "any" size

Well this is hard. Or not hard but resource intensive. What the preview code does now is round either the width or the height to the nearest power of two (round up). We do this so save processing power. And lets face it if a client reqests a 230x230 image. And they get a 256x256 image then they can scale that down just fine.

Also the current codes obeys the maxWidth and maxHeight settings for previews.

> obey aspect ratio

Well sure

> give correct landscape/portrait orientation

Again, sure.

> adjustable jpeg quality

That one I also find a bit tricky. Because you don't want your server to generate the same preview 5 times. Just because user1 has a quality of 90%, user2 of 85% etc.


So long story short. I think a dedicated endpoint is a good idea (webdav?). But lets not get to fancy with all the options. What you want it to request a preview of a given size and then get something fast. And not download a 2048x2048 image if all you wanted was a 100x100 image.

**Comment by tobiasKaminsky** [comment]
The options were just a quick thought, so I am fine with whatever I get ;-)
Rounding to the nearest preview is of course fine
Same for quality, some sane default should be enough 👍

**Comment by tobiasKaminsky** [comment]
> Yes, the Preview API should be its own app and come with a provider manager and an API

Would it then be shipped and enabled by default? Otherwise I would need something to check, e.g. in capabilities?

**Comment by oparoz** [comment]
>  Otherwise I would need something to check, e.g. in capabilities?

Probably the best way. 
Currently admins can completely disable previews and uninstall Gallery.

**Comment by tobiasKaminsky** [comment]
Easiest and "best" for me/user experience would be if it cannot be disabled ;-)

**Comment by rullzer** [comment]
This can be clsoed I believe.
